File: Algorithm/Programmers/level3_hash_베스트앨범.py
'''
TEST CASE

genres : ["classic", "pop", "classic", "classic", "pop"]	
plays : [500, 600, 150, 800, 2500]	

RESULT : [4, 1, 3, 0]

https://programmers.co.kr/learn/courses/30/lessons/42579
'''

def solution(genres, plays):
    answer = []
    length = len(genres)
    obj = {x : 0 for x in genres}
    
    for i in range(length) :
        obj[genres[i]] += plays[i]
    
    for q in obj :
        nowMax = max(obj.keys(),key=(lambda x:obj[x]))
        playsArr = [(j,plays[j]) for j in filter(lambda x:genres[x] == nowMax, range(length))]
        #for j in filter(lambda x:genres[x] == nowMax, range(length)) :
        #    playsArr.append((j,plays[j]))

        playsArr.sort(key=lambda x:x[1],reverse=True)
        obj[nowMax] = 0
        
        for k in playsArr[:2] :
            answer.append(k[0])
    
    return answer

File: Algorithm/Programmers/java/level3_hash_베스트앨범.java
/*
프로그래머스 hash level3

genres	plays	return
["classic", "pop", "classic", "classic", "pop"]	[500, 600, 150, 800, 2500]	[4, 1, 3, 0]

https://school.programmers.co.kr/learn/courses/30/lessons/42579
*/



import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> result = new ArrayList();
        Map<String,Integer> sumMap = new HashMap(); //합계
        Map<String,Map<Integer,Integer>> songMap = new HashMap(); //장르 - (index, number) Map
        
        for(int i=0; i < genres.length; i++){
            Map<Integer,Integer> idxMap = songMap.getOrDefault(genres[i], new HashMap());
            idxMap.put(i, plays[i]);
            songMap.put(genres[i], idxMap); //장르 Map
            sumMap.put(genres[i], sumMap.getOrDefault(genres[i], 0) + plays[i]); //합계
        }
        
        List<String> keyList = new ArrayList(sumMap.keySet());
        Collections.sort(keyList, (o1,o2) -> sumMap.get(o2) - sumMap.get(o1)); //합계 sort
        
        for(String key : keyList){
            Map<Integer,Integer> values = songMap.get(key);
            List<Integer> vList = new ArrayList(values.keySet());
            Collections.sort(vList, (o1,o2) -> values.get(o2) - values.get(o1)); //장르 Map sort
            
            result.add(vList.get(0));
            if(vList.size() > 1) result.add(vList.get(1));
        }
        
        return result.stream().mapToInt(i -> i).toArray();
    }
}
